source/source_backstage.go

```diff
@@ -119,6 +131,99 @@ func (s SourceBackstage) Load(ctx context.Context, logger kitlog.Logger, client
 	}
 }
 
+// https://backstage.io/docs/features/software-catalog/software-catalog-api/#get-entitiesby-query
+func (s SourceBackstage) fetchEntriesByQuery(ctx context.Context, client *http.Client, token string) ([]*SourceEntry, error) {
+	var (
+		limit  = defaultPageSize
+		cursor = ""
+	)
+
+	if s.PageSize != 0 {
+		limit = s.PageSize
+	}
+
+	entries := []*SourceEntry{}
+	for {
+		query := url.Values{}
+		query.Set("limit", strconv.Itoa(limit))
+		if cursor != "" {
+			query.Set("cursor", cursor)
+		}
+
+		if s.Filter != "" {
+			query.Set("filter", s.Filter)
+		}
+
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, s.Endpoint+"?"+query.Encode(), nil)
+		if err != nil {
+			return nil, errors.Wrap(err, "building Backstage URL")
+		}
+
+		if token != "" {
+
+			header := s.Header
+
+			if header == "" {
+				header = "Authorization"
+			}
+
+			req.Header.Add(header, fmt.Sprintf("Bearer %s", token))
+		}
+
+		resp, err := client.Do(req)
+		if err == nil && resp.StatusCode != http.StatusOK {
+			err = fmt.Errorf("received error from Backstage: %s", resp.Status)
+		}
+		if err != nil {
+			return nil, errors.Wrap(err, "fetching Backstage entries")
+		}
+
+		type GetEntitiesByQueryResponse struct {
```

Comment:
Could this be moved out of the function? It kinda broke the flow of reading and understanding the code for me